### ISSUE TITLE
azurerm_subnet resource - allow UDR linkage at creation time

### ISSUE THREAD

**Issue by molson504x** [hashicorp/terraform-provider-azurerm#17607]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Can the azurerm_subnet resource module be updated to support creating a subnet with a UDR (and NSG) attached?  I know that we can just create the subnet and then use the azurerm_subnet_route_table_association and azurerm_subnet_network_security_group_association resource module to do it immediately, but unfortunately our policy administrators have put a policy in place which prevents creating a subnet without a UDR associated so that procedure does not work for us.  Looking at the [`az network vnet subnet create` documentation](https://docs.microsoft.com/en-us/cli/azure/network/vnet/subnet?view=azure-cli-latest#az-network-vnet-subnet-create), this is supported which means the azure API also must support it ([azure API documentation here](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/subnets/create-or-update)).

### New or Affected Resource(s)/Data Source(s)

azurerm_subnet

### Potential Terraform Configuration

```hcl
Assumptions:
* This assumes that the NSG, UDR, Resource Group, and VNET already exist.

data "azurerm_resource_group" "rg" {
  name = "some-resource-group"
}

data "azurerm_virtual_network" "vnet" {
  name = "some-virtual-network"
  resource_group_name = data.azurerm_resource_group.rg.name
}

data "azurerm_network_security_group" "example_nsg" {
  name = "custom_nsg"
  resource_group_name = data.azurerm_resource_group.rg.name
}

data "azurerm_route_table" "example_udr" {
  name = "custom_route_table"
  resource_group_name = data.azurerm_resource_group.rg.name
}

resource "azurerm_subnet" "example" {
  name                 = "frontend"
  resource_group_name  = data.azurerm_resource_group.rg.name
  virtual_network_name = data.azurerm_virtual_network.vnet.name
  address_prefixes     = ["10.0.2.0/24"]
  route_table          = data.azurerm_route_table.example_udr.id
  network_security_group = data.azurerm_network_security_group.nsg.id
}
```

### References

* https://docs.microsoft.com/en-us/cli/azure/network/vnet/subnet?view=azure-cli-latest#az-network-vnet-subnet-create
* https://docs.microsoft.com/en-us/rest/api/virtualnetwork/subnets/create-or-update

**Comment by tombuildsstuff** [comment]
hi @molson504x 

These fields have been intentionally removed from the `azurerm_subnet` resource to workaround issues within the Azure Platform during the deletion of resources, as this field has been intentionally removed unfortunately this isn't something we plan to reintroduce.

Instead #9022 is tracking adding an example of how to use Azure Policy with the separate resources, and #3917 is tracking adding additional fields within the `azurerm_virtual_network` resource - as such whilst I'd like to thank you for opening this issue I'm going to close this in favour of #9022 - would you mind subscribing to that issue for updates?

Thanks!

**Comment by molson504x** [comment]
Hi @tombuildsstuff - thanks, I appreciate the reply.  Unfortuantely the solution proposed in #9022 wouldn't work for us because we do not create all of our subnets centrally - they are created on an as-needed basis.  What we do is have one RG in the subscription that houses all of the vnets for the subscription, and then each application being hosted in that subscription has its own resource group, etc...  We separate our terraform scripts out per environment (so, if an application has a dev and QA environment in our non-prod resource group, that's 2 separate sets of terraform scripts).  That's also how TFE workspaces would be managed too - each app environment would get its own TFE workspace.  So based off of that, our options are:
1. Create subnets manually using the portal (or az-cli or whatever) once terraform creates the UDR and NSG
2. Update our terraform modules to use a `local_exec` to execute az-cli commands.  At this point, there's no real point in using terraform.
3. Fundamentally change how terraform is being used within the organization - this is not really an option because of the size of the impact this would have.

**Comment by weisdd** [comment]
Hi @molson504x,
There was a recent discussion between me and @tombuildsstuff. Since the thread was marked off-topic and, thus, gets automatically hidden. I'll copy parts of it here, so other people would have a reference.

So, basically, we (the company I work for) are in a similar position to yours, only with NSG (subnet creation without an NSG attached is denied). In our case, virtual networks are centrally managed, and it's expected that users will create subnets within those virtual networks when needed.
`azurerm_subnet` resources cannot be managed through terraform, because NSG bindings is only possible via `azurerm_subnet_network_security_group_association`, so when terraform sends a request to create `azurerm_subnet` it gets already blocked by the policy.
End users don't have permissions to manage virtual networks, so `azurerm_virtual_network` with inline subnets is not possible in our scenario either.

I've searched through GitHub issues trying to understand what the motivation was to move the binding to a separate resource and found this one:
> "There's a bunch of reasons for this (mostly user experience related), but the main one is to work around a circular reference in Azure, where it's possible to create resources in any order; but on deletion you encounter an error where the Subnet deletion will fail unless the Network Security Group is detached first."
https://github.com/hashicorp/terraform-provider-azurerm/issues/3653#issuecomment-501995196

Since it all started years ago, I decided to make a PoC patch for azurerm provider to see if it's possible to delete a subnet without detaching NSG first: https://github.com/weisdd/terraform-provider-azurerm/commit/8ae1cdb66a43e341ebeee1d798c3168a9503e206, and it just worked, no issues at all:

```hcl
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_virtual_network" "example" {
  name                = "example-network"
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
}

resource "azurerm_subnet" "example" {
  name                 = "frontend"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes     = ["10.0.2.0/24"]
  security_group       = azurerm_network_security_group.example.id
}

resource "azurerm_network_security_group" "example" {
  name                = "example-nsg"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name

  security_rule {
    name                       = "test123"
    priority                   = 100
    direction                  = "Inbound"
    access                     = "Allow"
    protocol                   = "Tcp"
    source_port_range          = "*"
    destination_port_range     = "*"
    source_address_prefix      = "*"
    destination_address_prefix = "*"
  }
}
```

```shell
$ terraform apply
[...]
azurerm_resource_group.example: Creating...
azurerm_resource_group.example: Creation complete after 1s [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources]
azurerm_virtual_network.example: Creating...
azurerm_network_security_group.example: Creating...
azurerm_virtual_network.example: Creation complete after 4s [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources/providers/Microsoft.Network/virtualNetworks/example-network]
azurerm_network_security_group.example: Creation complete after 4s [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources/providers/Microsoft.Network/networkSecurityGroups/example-nsg]
azurerm_subnet.example: Creating...
azurerm_subnet.example: Creation complete after 4s [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources/providers/Microsoft.Network/virtualNetworks/example-network/subnets/frontend]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

$ terraform destroy
[...]

azurerm_subnet.example: Destroying... [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources/providers/Microsoft.Network/virtualNetworks/example-network/subnets/frontend]
azurerm_subnet.example: Still destroying... [id=/subscriptions/64842ced-4781-416f-81ff-...works/example-network/subnets/frontend, 10s elapsed]
azurerm_subnet.example: Destruction complete after 11s
azurerm_virtual_network.example: Destroying... [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources/providers/Microsoft.Network/virtualNetworks/example-network]
azurerm_network_security_group.example: Destroying... [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources/providers/Microsoft.Network/networkSecurityGroups/example-nsg]
azurerm_virtual_network.example: Still destroying... [id=/subscriptions/64842ced-4781-416f-81ff-...etwork/virtualNetworks/example-network, 10s elapsed]
azurerm_network_security_group.example: Still destroying... [id=/subscriptions/64842ced-4781-416f-81ff-...work/networkSecurityGroups/example-nsg, 10s elapsed]
azurerm_virtual_network.example: Destruction complete after 11s
azurerm_network_security_group.example: Destruction complete after 11s
azurerm_resource_group.example: Destroying... [id=/subscriptions/64842ced-4781-416f-81ff-482b7f562581/resourceGroups/example-resources]
azurerm_resource_group.example: Still destroying... [id=/subscriptions/64842ced-4781-416f-81ff-...62581/resourceGroups/example-resources, 10s elapsed]
azurerm_resource_group.example: Destruction complete after 15s

Destroy complete! Resources: 4 destroyed.
```

I reached out to Microsoft to double-check my observations and to make sure the behaviour will persist over time:
> (Question): "[...] I've tried to reintroduce NSG reference back to the azurerm_subnet resource (https://github.com/weisdd/terraform-provider-azurerm/commit/8ae1cdb66a43e341ebeee1d798c3168a9503e206) and found out that now it's actually possible to delete a subnet without detaching NSG, so Hashicorp's complaint seems to be no longer relevant. Could you track, when did that change happen in the Azure API backend? Will that behaviour persist over time? [...]"
> (Answer): "I confirm that the logic that allow subnet deletion without detaching NSG or UDR is permanent."

So, I asked @tombuildsstuff: 
> Given this information, would Hashicorp consider reintroducing the binding logic back to the subnet resource? :) I understand that you've been flooded with similar requests, though I haven't seen anyone saying that the Azure backend logic has got fixed at some point, thus everything is likely to work flawlessly now. I can prepare a PR with tests if needed.
https://github.com/hashicorp/terraform-provider-azurerm/issues/9022#issuecomment-1183195166

And got this reply:
>unfortunately reintroducing these fields to the azurerm_subnet resource would end up reintroduce the (hard to debug, but common in complex configurations) issues which caused this - so we wouldn't accept a PR to reintroduce those fields.
https://github.com/hashicorp/terraform-provider-azurerm/issues/9022#issuecomment-1183309333

So, even though it can be fixed for both NSG and UDR, it won't.
It overcomplicates our TFE experience, and we'll have to manually create subnets, then import their IDs either through manual copy-paste or through `data "azurerm_virtual_network"` blocks. We don't like any of those options.
Some teams in our company stopped using terraform for this reason.
